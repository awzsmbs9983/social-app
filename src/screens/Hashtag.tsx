import React from 'react'
import {ListRenderItemInfo, Pressable} from 'react-native'
import {PostView} from '@atproto/api/dist/client/types/app/bsky/feed/defs'
import {msg} from '@lingui/macro'
import {useLingui} from '@lingui/react'
import {useFocusEffect} from '@react-navigation/native'
import {NativeStackScreenProps} from '@react-navigation/native-stack'

import {HITSLOP_10} from 'lib/constants'
import {useInitialNumToRender} from 'lib/hooks/useInitialNumToRender'
import {CommonNavigatorParams} from 'lib/routes/types'
import {shareUrl} from 'lib/sharing'
import {cleanError} from 'lib/strings/errors'
import {sanitizeHandle} from 'lib/strings/handles'
import {enforceLen} from 'lib/strings/helpers'
import {isNative} from 'platform/detection'
import {useSearchPostsQuery} from 'state/queries/search-posts'
import {useSetMinimalShellMode} from 'state/shell'
import {Post} from 'view/com/post/Post'
import {List} from 'view/com/util/List'
import {ViewHeader} from 'view/com/util/ViewHeader'
import {ArrowOutOfBox_Stroke2_Corner0_Rounded} from '#/components/icons/ArrowOutOfBox'
import {
  ListFooter,
  ListHeaderDesktop,
  ListMaybePlaceholder,
} from '#/components/Lists'

const renderItem = ({item}: ListRenderItemInfo<PostView>) => {
  return <Post post={item} />
}

const keyExtractor = (item: PostView, index: number) => {
  return `${item.uri}-${index}`
}

export default function HashtagScreen({
  route,
}: NativeStackScreenProps<CommonNavigatorParams, 'Hashtag'>) {
  const {tag, author} = route.params
  const setMinimalShellMode = useSetMinimalShellMode()
  const {_} = useLingui()
  const initialNumToRender = useInitialNumToRender()
  const [isPTR, setIsPTR] = React.useState(false)

  const fullTag = React.useMemo(() => {
    return `#${decodeURIComponent(tag)}`
  }, [tag])

  const queryParam = React.useMemo(() => {
    if (!author) return fullTag
    return `${fullTag} from:${sanitizeHandle(author)}`
  }, [fullTag, author])

  const headerTitle = React.useMemo(() => {
    return enforceLen(fullTag.toLowerCase(), 24, true, 'middle')
  }, [fullTag])

  const sanitizedAuthor = React.useMemo(() => {
    if (!author) return
    return sanitizeHandle(author)
  }, [author])

  const {
    data,
    isFetchingNextPage,
    isLoading,
    isError,
    error,
    refetch,
    fetchNextPage,
    hasNextPage,
  } = useSearchPostsQuery({query: queryParam})

  const posts = React.useMemo(() => {
    return data?.pages.flatMap(page => page.posts) || []
  }, [data])

  useFocusEffect(
    React.useCallback(() => {
      setMinimalShellMode(false)
    }, [setMinimalShellMode]),
  )

  const onShare = React.useCallback(() => {
    const url = new URL('https://bsky.app')
    url.pathname = `/hashtag/${decodeURIComponent(tag)}`
    if (author) {
      url.searchParams.set('author', author)
    }
    shareUrl(url.toString())
  }, [tag, author])

  const onRefresh = React.useCallback(async () => {
    setIsPTR(true)
    await refetch()
    setIsPTR(false)
  }, [refetch])

  const onEndReached = React.useCallback(() => {
    if (isFetchingNextPage || !hasNextPage || error) return
    fetchNextPage()
  }, [isFetchingNextPage, hasNextPage, error, fetchNextPage])

  return (
    <>
      <ViewHeader
        title={headerTitle}
        subtitle={author ? _(msg`From @${sanitizedAuthor}`) : undefined}
        canGoBack
        renderButton={
          isNative
            ? () => (
                <Pressable
                  accessibilityRole="button"
                  onPress={onShare}
                  hitSlop={HITSLOP_10}>
                  <ArrowOutOfBox_Stroke2_Corner0_Rounded
                    size="lg"
                    onPress={onShare}
                  />
                </Pressable>
              )
            : undefined
        }
      />
      {posts.length < 1 ? (
        <ListMaybePlaceholder
          isLoading={isLoading}
          isError={isError}
          onRetry={refetch}
          emptyType="results"
          emptyMessage={_(msg`We couldn't find any results for that hashtag.`)}
        />
      ) : (
        <List
          data={posts}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
          refreshing={isPTR}
          onRefresh={onRefresh}
          onEndReached={onEndReached}
          onEndReachedThreshold={4}
          // @ts-ignore web only -prf
          desktopFixedHeight
          ListHeaderComponent={
            <ListHeaderDesktop
              title={headerTitle}
              subtitle={author ? _(msg`From @${sanitizedAuthor}`) : undefined}
            />
          }
          ListFooterComponent={
            <ListFooter
              isFetchingNextPage={isFetchingNextPage}
              error={cleanError(error)}
              onRetry={fetchNextPage}
            />
          }
          initialNumToRender={initialNumToRender}
          windowSize={11}
        />
      )}
    </>
  )
}
