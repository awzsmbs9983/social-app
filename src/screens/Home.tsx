import React from 'react'
import {Text, Button, View, SafeAreaView} from 'react-native'
import type {PrimaryTabScreenProps} from '../routes/types'

export const Home = ({navigation}: PrimaryTabScreenProps<'Home'>) => {
  return (
    <SafeAreaView style={{flex: 1}}>
      <View style={{flex: 1}}>
        <Text>Hello world</Text>
        <Button
          title="Go to Jane's profile"
          onPress={() => navigation.navigate('Profile', {name: 'Jane'})}
        />
      </View>
    </SafeAreaView>
  )
}
