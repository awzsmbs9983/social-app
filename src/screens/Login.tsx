import React from 'react'
import {Text, Button, View} from 'react-native'
import {Shell} from '../platform/shell'
import type {RootTabsScreenProps} from '../routes/types'
import {useStores} from '../state'

export function Login({navigation}: RootTabsScreenProps<'Login'>) {
  const store = useStores()
  return (
    <Shell>
      <View style={{justifyContent: 'center', alignItems: 'center'}}>
        <Text style={{fontSize: 20, fontWeight: 'bold'}}>Sign In</Text>
        <Button title="Login" onPress={() => store.session.setAuthed(true)} />
        <Button title="Sign Up" onPress={() => navigation.navigate('Signup')} />
      </View>
    </Shell>
  )
}
