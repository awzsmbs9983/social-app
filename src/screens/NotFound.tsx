import React from 'react'
import {Shell} from '../platform/shell'
import {Text, Button, View} from 'react-native'
import type {RootTabsScreenProps} from '../routes/types'

export const NotFound = ({navigation}: RootTabsScreenProps<'NotFound'>) => {
  return (
    <Shell>
      <View style={{justifyContent: 'center', alignItems: 'center'}}>
        <Text style={{fontSize: 20, fontWeight: 'bold'}}>Page not found</Text>
        <Button title="Home" onPress={() => navigation.navigate('Home')} />
      </View>
    </Shell>
  )
}
