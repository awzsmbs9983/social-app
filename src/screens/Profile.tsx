import React from 'react'
import {Text} from 'react-native'
import type {RootStackScreenProps} from '../routes/types'

export const Profile = ({route}: RootStackScreenProps<'Profile'>) => {
  return <Text>This is {route.params.name}'s profile</Text>
}
