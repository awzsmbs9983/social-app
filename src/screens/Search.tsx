import React from 'react'
import {SafeAreaView, ScrollView, Text, Button, View} from 'react-native'
import type {PrimaryTabScreenProps} from '../routes/types'

export const Search = ({navigation}: PrimaryTabScreenProps<'Search'>) => {
  return (
    <SafeAreaView>
      <ScrollView contentInsetAdjustmentBehavior="automatic">
        <View>
          <Text>Hello world</Text>
          <Button
            title="Go to Jane's profile"
            onPress={() => navigation.navigate('Profile', {name: 'Jane'})}
          />
        </View>
      </ScrollView>
    </SafeAreaView>
  )
}
