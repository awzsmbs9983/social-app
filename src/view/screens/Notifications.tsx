import React from 'react'
import {FlatList, View} from 'react-native'
import {useFocusEffect} from '@react-navigation/native'
import {observer} from 'mobx-react-lite'
import {
  NativeStackScreenProps,
  NotificationsTabNavigatorParams,
} from 'lib/routes/types'
import {withAuthRequired} from 'view/com/auth/withAuthRequired'
import {ViewHeader} from '../com/util/ViewHeader'
import {Feed} from '../com/notifications/Feed'
import {InvitedUsers} from '../com/notifications/InvitedUsers'
import {LoadLatestBtn} from 'view/com/util/LoadLatestBtn'
import {useStores} from 'state/index'
import {useOnMainScroll} from 'lib/hooks/useOnMainScroll'
import {s} from 'lib/styles'
import {useAnalytics} from 'lib/analytics'

type Props = NativeStackScreenProps<
  NotificationsTabNavigatorParams,
  'Notifications'
>
export const NotificationsScreen = withAuthRequired(
  observer(({}: Props) => {
    const store = useStores()
    const onMainScroll = useOnMainScroll(store)
    const scrollElRef = React.useRef<FlatList>(null)
    const {screen} = useAnalytics()

    // event handlers
    // =
    const onPressTryAgain = React.useCallback(() => {
      store.me.notifications.refresh()
    }, [store])

    const scrollToTop = React.useCallback(() => {
      scrollElRef.current?.scrollToOffset({offset: 0})
    }, [scrollElRef])

    const onPressLoadLatest = React.useCallback(() => {
      store.me.notifications.processQueue()
      scrollToTop()
    }, [store, scrollToTop])

    // on-visible setup
    // =
    useFocusEffect(
      React.useCallback(() => {
        store.shell.setMinimalShellMode(false)
        store.log.debug('NotificationsScreen: Updating feed')
        const softResetSub = store.onScreenSoftReset(onPressLoadLatest)
        store.me.notifications.syncQueue()
        store.me.notifications.update()
        screen('Notifications')

        return () => {
          softResetSub.remove()
          store.me.notifications.markAllUnqueuedRead()
        }
      }, [store, screen, onPressLoadLatest]),
    )

    return (
      <View testID="notificationsScreen" style={s.hContentRegion}>
        <ViewHeader title="Notifications" canGoBack={false} />
        <InvitedUsers />
        <Feed
          view={store.me.notifications}
          onPressTryAgain={onPressTryAgain}
          onScroll={onMainScroll}
          scrollElRef={scrollElRef}
        />
        {store.me.notifications.hasNewLatest &&
          !store.me.notifications.isRefreshing && (
            <LoadLatestBtn onPress={onPressLoadLatest} label="notifications" />
          )}
      </View>
    )
  }),
)
