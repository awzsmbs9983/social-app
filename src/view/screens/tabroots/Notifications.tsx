import React, {useState, useEffect, useLayoutEffect} from 'react'
import {Image, StyleSheet, TouchableOpacity, View} from 'react-native'
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome'
import {Shell} from '../../shell'
import {Feed} from '../../com/notifications/Feed'
import type {RootTabsScreenProps} from '../../routes/types'
import {useStores} from '../../../state'
import {AVIS} from '../../lib/assets'

export const Notifications = ({
  navigation,
}: RootTabsScreenProps<'NotificationsTab'>) => {
  const [hasSetup, setHasSetup] = useState<boolean>(false)
  const store = useStores()
  useEffect(() => {
    console.log('Fetching home feed')
    store.notesFeed.setup().then(() => setHasSetup(true))
  }, [store.notesFeed])

  const onNavigateContent = (screen: string, props: Record<string, string>) => {
    // @ts-ignore it's up to the callers to supply correct params -prf
    navigation.navigate(screen, props)
  }

  useEffect(() => {
    return navigation.addListener('focus', () => {
      if (hasSetup) {
        console.log('Updating home feed')
        store.notesFeed.update()
      }
    })
  }, [navigation, store.notesFeed, hasSetup])

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: true,
      headerTitle: 'Notifications',
      headerLeft: () => (
        <TouchableOpacity
          onPress={() => navigation.push('Profile', {name: 'alice.com'})}>
          <Image source={AVIS['alice.com']} style={styles.avi} />
        </TouchableOpacity>
      ),
      headerRight: () => (
        <TouchableOpacity
          onPress={() => {
            navigation.push('Composer', {})
          }}>
          <FontAwesomeIcon icon="plus" style={{color: '#006bf7'}} />
        </TouchableOpacity>
      ),
    })
  }, [navigation])

  return (
    <Shell>
      <View>
        <Feed view={store.notesFeed} onNavigateContent={onNavigateContent} />
      </View>
    </Shell>
  )
}

const styles = StyleSheet.create({
  avi: {
    width: 20,
    height: 20,
    borderRadius: 10,
    resizeMode: 'cover',
  },
})
