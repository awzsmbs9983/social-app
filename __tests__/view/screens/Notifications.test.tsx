import React from 'react'
import {Notifications} from '../../../src/view/screens/Notifications'
import renderer from 'react-test-renderer'
// import {render} from '../../../../jest/test-utils'

describe('Notifications', () => {
  const mockedProps = {
    navIdx: [0, 0] as [number, number],
    params: {},
    visible: true,
  }
  it('renders correctly', () => {
    const tree = renderer.create(<Notifications {...mockedProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
