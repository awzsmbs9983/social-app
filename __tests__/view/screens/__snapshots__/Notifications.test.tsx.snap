// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderBottomColor": "#f8f3f3",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "paddingBottom": 6,
        "paddingHorizontal": 12,
        "paddingTop": 6,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 30,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 30,
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={30}
        bbWidth={30}
        focusable={false}
        height={30}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "flex": 0,
              "height": 30,
              "width": 30,
            },
          ]
        }
        vbHeight={100}
        vbWidth={100}
        width={30}
      >
        <RNSVGGroup>
          <RNSVGDefs>
            <RNSVGLinearGradient
              gradient={
                Array [
                  0,
                  -1292135,
                  1,
                  -2424577,
                ]
              }
              gradientTransform={null}
              gradientUnits={0}
              name="grad"
              x1="0"
              x2="1"
              y1="0"
              y2="1"
            />
          </RNSVGDefs>
          <RNSVGCircle
            cx="50"
            cy="50"
            fill={
              Array [
                1,
                "grad",
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            r="50"
          />
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={4294967295}
            font={
              Object {
                "fontSize": "50",
                "fontWeight": "bold",
                "textAnchor": "middle",
              }
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                "50",
              ]
            }
            y={
              Array [
                "67",
              ]
            }
          >
            <RNSVGTSpan
              content="X"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
    <View
      pointerEvents="none"
      style={
        Object {
          "alignItems": "baseline",
          "flexDirection": "row",
          "marginRight": "auto",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
            },
            Object {
              "color": "#000000",
              "fontSize": 21,
              "fontWeight": "600",
            },
          ]
        }
      >
        Notifications
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f8f3f3",
          "borderRadius": 20,
          "flexDirection": "row",
          "height": 36,
          "justifyContent": "center",
          "opacity": 1,
          "width": 36,
        }
      }
    >
      <
        icon="plus"
        size={18}
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f8f3f3",
          "borderRadius": 20,
          "flexDirection": "row",
          "height": 36,
          "justifyContent": "center",
          "marginLeft": 8,
          "opacity": 1,
          "width": 36,
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={18}
        bbWidth={18}
        color={4278190080}
        fill="none"
        focusable={false}
        height={18}
        meetOrSlice={0}
        minX={0}
        minY={0}
        stroke="currentColor"
        strokeWidth={3}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "color": "#000000",
              "position": "relative",
              "top": -1,
            },
            Object {
              "flex": 0,
              "height": 18,
              "width": 18,
            },
          ]
        }
        tintColor={4278190080}
        vbHeight={24}
        vbWidth={24}
        width={18}
      >
        <RNSVGGroup
          fill={null}
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={
            Array [
              2,
            ]
          }
          strokeWidth={3}
        >
          <RNSVGPath
            d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
            propList={
              Array [
                "strokeLinecap",
                "strokeLinejoin",
              ]
            }
            strokeLinecap={1}
            strokeLinejoin={1}
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderRadius": 20,
          "flexDirection": "row",
          "height": 36,
          "justifyContent": "center",
          "marginLeft": 8,
          "opacity": 1,
          "width": 36,
        }
      }
    >
      <
        icon="signal"
        size={18}
        style={
          Array [
            Object {
              "color": "#000000",
            },
          ]
        }
      />
      <
        icon="x"
        size={12}
        style={
          Object {
            "backgroundColor": "#ffffff",
            "color": "#d1106f",
            "left": -4,
            "position": "relative",
            "top": 6,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
